rfd/0180-sso-as-mfa-method.md

```diff
@@ -0,0 +1,425 @@
+> tsh ssh server01
+Re-authentication is required to access this node.
+Complete an auth check in your local web browser. Open it by clicking on the link:
+ http://127.0.0.1:60433/f5858c78-75e1-4f3f-b2c5-69e8e76c0ff9
+Or tap any security key
+```
+
+### Configuration
+
+#### Enable SSO connector as MFA method.
+
+IdP connectors will gain a new `allow_as_mfa_method` field with possible values
+`no`, `yes`, and `only`. 
+
+When set to `no`, the connector cannot be used for MFA checks, and vice versa
+for `yes`.
+
+When set to `only`, the connector can be used for MFA checks but cannot be used
+for login. This option is useful in cases where administrators want to set up
+a second IdP which performs a subset of the login checks. e.g. just Webauthn
+without password.
+
+Note: a split Login/MFA IdP setup requires that both IdPs are set up for the
+same list of users. If a user tries to authenticate through the MFA IdP, and
+the resulting login's username doesn't match, it will result in an error. The
+user will be prompted to add an MFA device to give them a path forward while
+administrators sort out the misconfiguration.
+
+```yaml
+kind: saml
+version: v2
+metadata:
+  name: okta
+spec:
+  allow_as_mfa_method: no | yes | only
+```
+
+#### Make SSO the only MFA method
+
+Using SSO as an MFA method enables Administrators to maintain tighter control
+over what MFA devices can be used for MFA. In some cases, it may make sense to
+disable non-SSO MFA methods to prevent users from going around registered SSO
+MFA connectors.
+
+For this use case, we will add `second_factor: sso`, which will prevent users
+from registering/using MFA devices registered through Teleport.
+
+#### Default MFA connector
+
+Cluster auth preference will also gain the `mfa_connector_name` field to set
+a preferred IdP connector for MFA checks. As detailed in the UX section, setting
+a preferred IdP connector has some beneficial UX implications.
+
+If `mfa_connector_name` is not set, but `connector_name` is set, that connector
+will be used as an MFA method if `allow_as_mfa_method` is set to `yes`.
+
+If neither of the fields above are set, Teleport will look through all
+registered connectors and return the first one with `allow_as_mfa_method` set
+to `yes` or`only`.
+
+#### Bad configuration
+
+The following two fail states should be prevented:
+- `mfa_connector_name` points to a connector with `allow_as_mfa_method = no`
+- `connector_name` points to a connector with `allow_as_mfa_method = only`
+
+These fail states will be checked on both connector update and auth preference
+update.
+
+### Security
+
+Teleport uses MFA checks for some of its most security focused features, including
+per-session MFA, moderated sessions, and MFA for admin actions. Using SSO as an
+MFA method opens up the possibility of poorly configured clusters being
+vulnerable to attacks ranging from internal users avoiding safe MFA checks to
+attackers with a compromised IdP gaining keys to the castle.
+
```

Comment:
How about locking? Today we can lock an MFA device by ID - will we have a way to place a lock on SSO MFA? If so, would it apply for all users or would we be able to only lock it for a specific user?

Reply:
Yes, locking the MFA device should work since we keep a backend resource for it. The only issue is that it's an ephemeral resource, so we'll need to prevent its deletion if a lock exists, if that logic isn't already in place.

To put a "lock" on an SSO MFA method across the board, deleting the auth connector should work, or implementing auth connector locking.

Reply:
Add a locking session to the RFD?

Could we specify the lock in a way that says "lock the device from connector X", so it doesn't depend on specifics of the ephemeral device? (This is a bit worse than locking by the ID, but I think it's fine for a corner case like this.)

Reply:
In the updated design, we don't actually store an ephemeral device, so there would be no MFA device UUID to lock. If we tried to lock by connector name, the lock would apply to all users. We could use the user's UUID as the device UUID but that's a bit contrived.

Do we need locking on SSO MFA "devices"? If there is a misconfiguration or security issue with a user's IdP config, wouldn't we just want to lock the user completely?

Reply:
We could create a "device" lock for (user, connector), although this kind of lock doesn't exist today.

> Do we need locking on SSO MFA "devices"?

That's a good question. I think we can punt until there's a clear need (if ever?), but still it's worth a mention in the RFD.